<?xml version="1.0"?>
<launch>

	<!-- Parameters -->
	<arg name="use_gui" default="true" />
	<arg name="use_rviz" default="true" />
	<arg name="world_name" default="labyrinth.world" />

	<!--if	true it starts lidar-->
	<arg name="sensors" default="true" />
	<arg name="spawn_x" default="0.0" />
	<arg name="spawn_y" default="0." />
	<arg name="spawn_z" default="0.2" />
	<arg name="spawn_Y" default="0.0" />

	<arg name="start_controller" value="true" />

	<arg name="robot_id" value="0" />

	<!-- Set parameters -->
	<param name="use_sim_time" value="true" />
	<env name="GAZEBO_MODEL_PATH" value="$(find map_pkg)/models" />
	<env name="GAZEBO_RESOURCE_PATH" value="$(find map_pkg)/worlds" />

	<!--Start 	gazebo-->
	<include file="$(find gazebo_ros)/launch/empty_world.launch">
		<arg name="world_name" value="$(find map_pkg)/worlds/$(arg world_name)" />
		<arg name="paused" value="false" />
		<arg name="use_sim_time" value="true" />
		<arg name="gui" value="$(arg use_gui)" />
		<arg name="headless" value="false" />
		<arg name="debug" value="false" />
		<arg name="verbose" value="true" />
	</include>

	<!--MAP	SERVER: load recorded map for amcl-->
	<node name="map_server" pkg="map_server" type="map_server" args="$(find map_pkg)/maps/labyrinth.yaml" />

	<group ns="limo$(arg robot_id)">

		<!-- Description with per-robot namespace -->
		<include file="$(find limo_description)/launch/upload.launch">
			<arg name="robot_id" default="$(arg robot_id)" />
			<arg name="sensors" default="$(arg sensors)" />
		</include>

	    <!-- Spawn limo0 in simulation -->
	    <node name="spawn_model" pkg="gazebo_ros" type="spawn_model"   args="-urdf -param robot_description -model limo$(arg robot_id) -param robot_description -model limo0 -x $(arg spawn_x) -y $(arg spawn_y) -z $(arg spawn_z)  -Y $(arg spawn_Y)" />

		<!-- TF & joints -->
		<node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher">
			<param name="rate" value="100" />
		</node>

		<!--AMCL: publish map -> odom TF -->
		<node pkg="amcl" type="amcl" name="amcl">
			<!-- Fix: remap static_map request to global -->
			<remap from="static_map" to="/static_map" />
			<param name="odom_frame_id" value="limo$(arg robot_id)/odom" />
			<param name="base_frame_id" value="limo$(arg robot_id)/base_link" />
			<param name="global_frame_id" value="map" />
			<!-- Initial pose for limo0 -->
			<param name="initial_pose_x" value="$(arg spawn_x)" />
			<param name="initial_pose_y" value="$(arg spawn_y)" />
			<param name="initial_pose_a" value="$(arg spawn_Y)" />
		</node>

		<group if="$(arg start_controller)">
			<node pkg="loco_planning" type="controller.py" name="controller" output="screen"
				args="--robot_id  $(arg robot_id)  --debug false" />
		</group>
	</group>


	<!-- PUBLISH OBSTACLES RETRIEVED FROM IMAGE CONTOURS and publish them-->
	<node pkg="map_pkg" type="convert_labyrinth_map.py" name="map_extract_contours_into_obstacles" output="screen" />

	<!-- RViz (optional) -->
	<node pkg="rviz" type="rviz" name="rviz" if="$(arg use_rviz)"
		args="-d $(find loco_planning)/rviz/conf_localize.rviz" />

</launch>
